### PR TITLE
Added Typescript declarations.

### PR DESCRIPTION
**Pull request by jensvdh**
Hi,

I added Typescript definitions to this repository. This should help Typescript developers a lot by providing type declarations for this repository.

See issue #256 

Big thanks to @RichiWIP whom provided the first version of a Typescript declaration file.

I decided to follow Microsoft's recommended new guidelines that suggest bundling the Typescript Declarations with the module and adding a reference to package.json. Therefore developers won't have to download the typescript definitions from an external repository.

see:
https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html

Alternatively we can create a second repository and submit our typings to Microsoft's @types project directly. However, this process is more involved and no longer seems recommended.

I added some steps to readme.md to get Typescript developers started. Let me know if you'd like to put this documentation somewhere else.

### CODE REVIEW

**Comment by WolframHempel** [comment]
Awesome, thank you very much. We'll review it and merge it shortly

File: src/client.d.ts [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+        Info
+        
+        Requests count as completed once send() or error() was called. Calling ack() after that won't do anything.
+                 */
+                ack(): void;
+        }
+
+        interface RecordStatic {
+                /** Get a record. */
+                getRecord(path: string): Record;
```
Comment (yasserf) [inline src/client.d.ts]: getAnonymousRecord is missing here ( as well as an interface )

**Comment by yasserf** [comment]
Any chance anyone tested or used this? Be great to get in for 2.1

**Comment by jensvdh** [comment]
What's the schedule for 2.1? We noticed a couple of issues  that need to be addressed with the typings.

**Comment by yasserf** [comment]
We are aiming to get it out by early next week. No rush though, we can easily just merge it in when your happy and do a separate release, our release cycle is mostly automated ;) =D

**Comment by NathanFlurry** [comment]
I started trying out this declaration file and immediately ran into a few issues. First of all, none of the interfaces or enums are exported, so it's impossible to store values like the `Client` in TypeScript as a non-`any` type. Additionally, TypeScript enums are not the same as string enums. For instance, `ConnectionState.OPEN` would equal `5` based on the declaration file, but in the JavaScript library, it equals the literal string, `"OPEN"`. Therefore, I believe it needs a bit more work. If I get a chance, I'll look into fixing it.

**Comment by jensvdh** [comment]
I believe you can use strings in ENUM's now in Typescript like this:
https://blog.rsuter.com/how-to-implement-an-enum-with-string-values-in-typescript/

**Comment by yasserf** [comment]
I'm sorry for the delay on merging this, we are still trying to figure out the best way of testing and maintaining it going forward. It will be merged, just pending the next client milestone.

File: src/client.d.ts [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+
+                rpc: RPCStatic;
+        }
+
+        interface deepstreamQuarantine extends Client {
+
+        }
+        interface deepstreamStatic {
+                CONSTANTS: {
+                        [key: string]: string
```
Comment (jensvdh) [inline src/client.d.ts]: That works. We'll also want to update the enums to use strings. Typescript enums compile to integers.

**Comment by yasserf** [comment]
@jensvdh @benjaminVadon Are you both okay with me merging this?

**Comment by jensvdh** [comment]
I'm OK with merging this. It's not perfect yet but we can improve on it in a 2nd PR. Have these typings is still an improvement over having none.

**Comment by yasserf** [comment]
Agreed!

**Comment by yasserf** [comment]
One last thing, would you be okay with signing the CLA? https://deepstream.io/info/community/cla/

I have said that a bit often lately so if you already have my bad!

**Comment by jensvdh** [comment]
Done!

**Comment by yasserf** [comment]
Awesome, and thanks again for raising this!

**Pull request merged**